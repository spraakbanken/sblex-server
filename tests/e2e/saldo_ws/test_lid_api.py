import xml.etree.ElementTree as ET

import pytest
from fastapi import status
from httpx import AsyncClient

EXPECTED_XML_RESPONSES = {
    "empty": '<?xml version="1.0" encoding="UTF-8"?>\n<result>\n\n</result>',
    "dväljas..vb.1": """<result>
<gf>dväljas</gf>
    <p>vb_vs_dväljas</p>
    <ls>
    <l>dväljas..1</l>
    </ls>

</result>
""",
    "dväljas..1": """<result>
<lex>dväljas..1</lex>
    <fm>bo..1</fm>
    <fp>PRIM..1</fp>
    <mfs></mfs>
    <pfs></pfs>
    <ls>
    <l>dväljas..vb.1</l>
    </ls>
</result>
""",
}

EXPECTED_PROTOJS_RESPONSES = {
    "bo..1": """var flare = {
 'dväljas..1': 'http://testserver/lid/graph/dväljas..1', 
}"""
}


class TestLidRoutes:
    @pytest.mark.parametrize(
        "in_format",
        [
            "json",
            "xml",
            "html",
        ],
    )
    @pytest.mark.asyncio
    async def test_invalid_input_returns_422(self, client: AsyncClient, in_format: str) -> None:
        res = await client.get(f"/lid/{in_format}/bad-input")
        assert res.status_code == status.HTTP_422_UNPROCESSABLE_ENTITY

    @pytest.mark.parametrize(
        "lid, expected_response",
        [
            (
                "dväljas..vb.1",
                {"gf": "dväljas", "l": ["dväljas..1"], "p": "vb_vs_dväljas"},
            ),
            (
                "dväljas..1",
                {
                    "fm": "bo..1",
                    "fp": "PRIM..1",
                    "l": ["dväljas..vb.1"],
                    "lex": "dväljas..1",
                    "mf": [],
                    "path": ["bo..1"],
                    "pf": [],
                    "ppath": [],
                },
            ),
            # ("d..nn.1", {}),
        ],
    )
    @pytest.mark.asyncio
    async def test_json_valid_input_returns_200(
        self,
        client: AsyncClient,
        lid: str,
        expected_response: list,
    ) -> None:
        res = await client.get(f"/lid/json/{lid}")
        assert res.status_code == status.HTTP_200_OK
        assert res.headers["content-type"] == "application/json"

        assert res.json() == expected_response

    @pytest.mark.parametrize(
        "lid, expected_response_name",
        [
            (
                "xxxxx..xx.1",
                "empty",
            ),
            ("dväljas..vb.1", "dväljas..vb.1"),
            (
                "xxxxx..1",
                "empty",
            ),
            (
                "dväljas..1",
                "dväljas..1",
            ),
        ],
    )
    @pytest.mark.asyncio
    async def test_xml_valid_input_returns_200(
        self,
        client: AsyncClient,
        lid: str,
        expected_response_name: str,
    ) -> None:
        res = await client.get(f"/lid/xml/{lid}")
        assert res.status_code == status.HTTP_200_OK
        assert res.headers["content-type"] == "application/xml"

        expected_xml = ET.canonicalize(EXPECTED_XML_RESPONSES[expected_response_name])

        response_xml = ET.canonicalize(res.text)
        assert response_xml == expected_xml

    @pytest.mark.parametrize(
        "lid, expected_in_response",
        [
            # ("dväljs..vb.1", ["<h1>dväljas</h1>"]),
            ("xxxxx..xx.1", ["<b>xxxxx..xx.1 finns ej.</b>"]),
        ],
    )
    @pytest.mark.asyncio
    async def test_html_valid_lemma_returns_200(
        self,
        client: AsyncClient,
        lid: str,
        expected_in_response: list[str],
    ) -> None:
        res = await client.get(f"/lid/html/{lid}")
        assert res.status_code == status.HTTP_200_OK
        assert res.headers["content-type"] == "text/html; charset=utf-8"
        print(f"{res.text=}")
        for expect_in_response in expected_in_response:
            assert expect_in_response in res.text

    @pytest.mark.parametrize(
        "lid, expected_in_response",
        [
            ("dväljas..1", ["<h1>dväljas</h1>", '<td><a href="', "lid/html/bo..1"]),
            ("xxxxx..1", ["<center>xxxxx..1 finns ej.</center>"]),
        ],
    )
    @pytest.mark.asyncio
    async def test_html_valid_lexeme_returns_200(
        self,
        client: AsyncClient,
        lid: str,
        expected_in_response: list[str],
    ) -> None:
        res = await client.get(f"/lid/html/{lid}")
        assert res.status_code == status.HTTP_200_OK
        assert res.headers["content-type"] == "text/html; charset=utf-8"
        print(f"{res.text=}")
        for expect_in_response in expected_in_response:
            assert expect_in_response in res.text

    @pytest.mark.parametrize(
        "lid",
        [
            "dväljas..vb.1",
        ],
    )
    @pytest.mark.parametrize(
        "in_format, expected_content_type",
        [
            ("json", "application/json"),
            ("xml", "application/xml"),
            ("html", "text/html; charset=utf-8"),
        ],
    )
    @pytest.mark.asyncio
    async def test_valid_input_returns_200(
        self,
        client: AsyncClient,
        in_format: str,
        expected_content_type: str,
        lid: str,
    ) -> None:
        print(f"{lid=}")
        res = await client.get(f"/lid/{in_format}/{lid}")
        assert res.status_code == status.HTTP_200_OK
        assert res.headers["content-type"] == expected_content_type

    @pytest.mark.parametrize(
        "lid",
        [
            "dväljas..vb.1",
            "bad-input",
        ],
    )
    @pytest.mark.asyncio
    async def test_protojs_non_lexeme_input_returns_422(
        self,
        client: AsyncClient,
        lid: str,
    ) -> None:
        print(f"{lid=}")
        res = await client.get(f"/lid/protojs/{lid}")
        assert res.status_code == status.HTTP_422_UNPROCESSABLE_ENTITY

    @pytest.mark.parametrize(
        "lid, expected_response_name",
        [
            ("bo..1", None),
        ],
    )
    @pytest.mark.asyncio
    async def test_protojs_valid_input_returns_200(
        self,
        client: AsyncClient,
        expected_response_name: str | None,
        lid: str,
    ) -> None:
        print(f"{lid=}")
        res = await client.get(f"/lid/protojs/{lid}")
        assert res.status_code == status.HTTP_200_OK
        assert res.headers["content-type"] == "text/javascript; charset=utf-8"

        expected_response = EXPECTED_PROTOJS_RESPONSES[expected_response_name or lid]

        assert res.text == expected_response

    @pytest.mark.parametrize(
        "lid",
        [
            "dväljas..vb.1",
            "bad-input",
        ],
    )
    @pytest.mark.asyncio
    async def test_graph_non_lexeme_input_returns_422(
        self,
        client: AsyncClient,
        lid: str,
    ) -> None:
        res = await client.get(f"/lid/graph/{lid}")
        assert res.status_code == status.HTTP_422_UNPROCESSABLE_ENTITY

    @pytest.mark.parametrize(
        "lid, expected_in_response",
        [
            (
                "bo..1",
                '<script type="text/javascript" src="http://testserver/lid/protojs/bo..1"></script>',
            ),
        ],
    )
    @pytest.mark.asyncio
    async def test_graph_valid_input_returns_200(
        self,
        client: AsyncClient,
        expected_in_response: str,
        lid: str,
    ) -> None:
        res = await client.get(f"/lid/graph/{lid}")
        assert res.status_code == status.HTTP_200_OK
        assert res.headers["content-type"] == "text/html; charset=utf-8"

        assert expected_in_response in res.text
