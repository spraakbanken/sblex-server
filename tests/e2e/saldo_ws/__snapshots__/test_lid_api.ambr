# serializer version: 1
# name: TestLidRoutes.test_graph_valid_input_returns_200[bo..1]
  '''
  <!doctype html>
  <html lang="sv">
      <head>
          <meta content="text/html; charset=utf-8" http-equiv="Content-Type" />
          <link
              rel="shortcut icon"
              href="https://svn.spraakdata.gu.se/sb-arkiv/pub/images/saldo_icon.png"
          />
          <link
              rel="stylesheet"
              type="text/css"
              href="http://testserver/static/saldo.css"
          />
          <title>Lemma-id | graf | bo..1</title>
          
          <script
              src="https://unpkg.com/htmx.org@1.9.8"
              crossorigin="anonymous"
          ></script>
      </head>
  
      <body>
          <center>
              <p>
                  <a href="http://testserver/fl/html"
                      ><img
                          src="https://spraakbanken.gu.se/sites/spraakbanken.gu.se/files/img/saldo/saldo.gif"
                          align="top"
                          alt="SALDO: gå tillbaka till startsidan"
                  /></a>
              </p>
          </center>
          <div id="output_table">
  <div style="width: 500px; height: 500px;">
      <script type="text/javascript" src="https://svn.spraakdata.gu.se/repos/sblex/pub/js/protovis-r3.2.js"></script>
      <script type="text/javascript" src="http://testserver/lid/protojs/bo..1"></script>
      <script type="text/javascript+protovis">
  var vis = new pv.Panel()
      .width(500)
      .height(500);
  
  var tree = vis.add(pv.Layout.Tree)
      .nodes(pv.dom(flare).root("bo..1").nodes())
      .depth(150)
      .breadth(100)
      .orient("radial");
  
  tree.link.add(pv.Line);
  tree.node.add(pv.Dot)
      .fillStyle(function(n) n.firstChild ? "#ffffff" : "#ffffff")
      .size(10)
      .event("mouseover", function() this.fillStyle("blue"))
      .event("mouseout", function() this.fillStyle(undefined))
      .event("click", function(n) {
      if(n.nodeValue != undefined){
      document.location.href = n.nodeValue;
      }
      else{
      
      document.location.href = 'http://testserver/lid/graph/leva..1';
      
      }
      });
  tree.label.add(pv.Label)
      .text(function(n) n.nodeName.replace('..1','').replace('..','_'))
      .font("12px 'Arvo', sans-serif")
      .events("all")
      .cursor("hand")
      .event("click", function(n) {
      if(n.nodeValue != undefined){
      document.location.href = n.nodeValue;
         }
       else{
          
          document.location.href = 'http://testserver/lid/graph/leva..1';
          
       }
       });
   
   
   vis.render();
      </script>
  </div>
  </div>
          <footer>
              <hr />
              <ul>
                  <li>
                      <a href="https://spraakbanken.gu.se/saldo"
                          >SALDOs projektsida</a
                      >
                  </li>
                  <li>
                      <a
                          href="https://svn.spraakdata.gu.se/repos/sblex/pub/saldo_instruktion.pdf"
                          >dokumentation (pdf)</a
                      >
                  </li>
              </ul>
          </footer>
      </body>
  </html>
  '''
# ---
# name: TestLidRoutes.test_html_missing_lemma_returns_404[xxxxx..xx.1]
  '''
  <!doctype html>
  <html lang="sv">
      <head>
          <meta content="text/html; charset=utf-8" http-equiv="Content-Type" />
          <link
              rel="shortcut icon"
              href="https://svn.spraakdata.gu.se/sb-arkiv/pub/images/saldo_icon.png"
          />
          <link
              rel="stylesheet"
              type="text/css"
              href="http://testserver/static/saldo.css"
          />
          <title>Lemma-id saknas</title>
          
          <script
              src="https://unpkg.com/htmx.org@1.9.8"
              crossorigin="anonymous"
          ></script>
      </head>
  
      <body>
          <center>
              <p>
                  <a href="http://testserver/fl/html"
                      ><img
                          src="https://spraakbanken.gu.se/sites/spraakbanken.gu.se/files/img/saldo/saldo.gif"
                          align="top"
                          alt="SALDO: gå tillbaka till startsidan"
                  /></a>
              </p>
          </center>
          <div id="output_table">
  <center><p><b>xxxxx..xx.1 finns ej.</b></p></center>
  </div>
          <footer>
              <hr />
              <ul>
                  <li>
                      <a href="https://spraakbanken.gu.se/saldo"
                          >SALDOs projektsida</a
                      >
                  </li>
                  <li>
                      <a
                          href="https://svn.spraakdata.gu.se/repos/sblex/pub/saldo_instruktion.pdf"
                          >dokumentation (pdf)</a
                      >
                  </li>
              </ul>
          </footer>
      </body>
  </html>
  '''
# ---
# name: TestLidRoutes.test_html_missing_lexeme_returns_404[xxxxx..1]
  '''
  <!doctype html>
  <html lang="sv">
      <head>
          <meta content="text/html; charset=utf-8" http-equiv="Content-Type" />
          <link
              rel="shortcut icon"
              href="https://svn.spraakdata.gu.se/sb-arkiv/pub/images/saldo_icon.png"
          />
          <link
              rel="stylesheet"
              type="text/css"
              href="http://testserver/static/saldo.css"
          />
          <title>Saldo-id saknas</title>
          
          <script
              src="https://unpkg.com/htmx.org@1.9.8"
              crossorigin="anonymous"
          ></script>
      </head>
  
      <body>
          <center>
              <p>
                  <a href="http://testserver/fl/html"
                      ><img
                          src="https://spraakbanken.gu.se/sites/spraakbanken.gu.se/files/img/saldo/saldo.gif"
                          align="top"
                          alt="SALDO: gå tillbaka till startsidan"
                  /></a>
              </p>
          </center>
          <div id="output_table">
  <center>xxxxx..1 finns ej.</center>
  </div>
          <footer>
              <hr />
              <ul>
                  <li>
                      <a href="https://spraakbanken.gu.se/saldo"
                          >SALDOs projektsida</a
                      >
                  </li>
                  <li>
                      <a
                          href="https://svn.spraakdata.gu.se/repos/sblex/pub/saldo_instruktion.pdf"
                          >dokumentation (pdf)</a
                      >
                  </li>
              </ul>
          </footer>
      </body>
  </html>
  '''
# ---
# name: TestLidRoutes.test_html_valid_lemma_returns_200[dv\xe4ljas..vb.1]
  '''
  <!doctype html>
  <html lang="sv">
      <head>
          <meta content="text/html; charset=utf-8" http-equiv="Content-Type" />
          <link
              rel="shortcut icon"
              href="https://svn.spraakdata.gu.se/sb-arkiv/pub/images/saldo_icon.png"
          />
          <link
              rel="stylesheet"
              type="text/css"
              href="http://testserver/static/saldo.css"
          />
          <title>Lemma-id | dväljas..vb.1</title>
          
          <script
              src="https://unpkg.com/htmx.org@1.9.8"
              crossorigin="anonymous"
          ></script>
      </head>
  
      <body>
          <center>
              <p>
                  <a href="http://testserver/fl/html"
                      ><img
                          src="https://spraakbanken.gu.se/sites/spraakbanken.gu.se/files/img/saldo/saldo.gif"
                          align="top"
                          alt="SALDO: gå tillbaka till startsidan"
                  /></a>
              </p>
          </center>
          <div id="output_table">
  </div>
          <footer>
              <hr />
              <ul>
                  <li>
                      <a href="https://spraakbanken.gu.se/saldo"
                          >SALDOs projektsida</a
                      >
                  </li>
                  <li>
                      <a
                          href="https://svn.spraakdata.gu.se/repos/sblex/pub/saldo_instruktion.pdf"
                          >dokumentation (pdf)</a
                      >
                  </li>
              </ul>
          </footer>
      </body>
  </html>
  '''
# ---
# name: TestLidRoutes.test_html_valid_lexeme_returns_200[dv\xe4ljas..1]
  '''
  <!doctype html>
  <html lang="sv">
      <head>
          <meta content="text/html; charset=utf-8" http-equiv="Content-Type" />
          <link
              rel="shortcut icon"
              href="https://svn.spraakdata.gu.se/sb-arkiv/pub/images/saldo_icon.png"
          />
          <link
              rel="stylesheet"
              type="text/css"
              href="http://testserver/static/saldo.css"
          />
          <title>Saldo-id | dväljas..1</title>
          
          <script
              src="https://unpkg.com/htmx.org@1.9.8"
              crossorigin="anonymous"
          ></script>
      </head>
  
      <body>
          <center>
              <p>
                  <a href="http://testserver/fl/html"
                      ><img
                          src="https://spraakbanken.gu.se/sites/spraakbanken.gu.se/files/img/saldo/saldo.gif"
                          align="top"
                          alt="SALDO: gå tillbaka till startsidan"
                  /></a>
              </p>
          </center>
          <div id="output_table"><center>
      <h1>dväljas</h1>
  </center>
  <center>
      <table border="1">
          <tr>
              <td style="text-align:center;">⇧[2]</td>
              <td><a href="http://testserver/lid/html/bo..1">bo</a></td>
              <td style="text-align:center;">↑</td>
              <td><a href="http://testserver/lid/html/PRIM..1">PRIM</a></td>
          </tr>
      </table>
      <p><a href="http://testserver/lid/html/dväljas..vb.1">dväljas (vb)</a><br /> <a href="http://testserver/lid/graph/dväljas..1"><img
          src="https://svn.spraakdata.gu.se/repos/sblex/pub/images/prim_graph.png" /></a> <br /> <a href="http://spraakbanken.gu.se/korp/#search=lemgram|dväljas..vb.1">[korpus]</a></p>
      <table border="1">
          <tr>
              <td style="vertical-align:top;text-align:center;">⇩</td>
              <td style="vertical-align:top;">*</td>
              <td style="vertical-align:top;text-align:center;">↓</td>
              <td style="vertical-align:top;">*</td>
          </tr>
      </table>
  </center></div>
          <footer>
              <hr />
              <ul>
                  <li>
                      <a href="https://spraakbanken.gu.se/saldo"
                          >SALDOs projektsida</a
                      >
                  </li>
                  <li>
                      <a
                          href="https://svn.spraakdata.gu.se/repos/sblex/pub/saldo_instruktion.pdf"
                          >dokumentation (pdf)</a
                      >
                  </li>
              </ul>
          </footer>
      </body>
  </html>
  '''
# ---
# name: TestLidRoutes.test_protojs_valid_input_returns_200[bo..1]
  '''
  var flare = {
   'dväljas..1': 'http://testserver/lid/graph/dväljas..1', 
  }
  '''
# ---
# name: TestLidRoutes.test_xml_missing_returns_404[xxxxx..1]
  '''
  <?xml version="1.0" encoding="UTF-8"?>
  <result>
  
  </result>
  '''
# ---
# name: TestLidRoutes.test_xml_missing_returns_404[xxxxx..xx.1]
  '''
  <?xml version="1.0" encoding="UTF-8"?>
  <result>
  
  </result>
  '''
# ---
# name: TestLidRoutes.test_xml_valid_input_returns_200[dv\xe4ljas..1]
  '''
  <?xml version="1.0" encoding="UTF-8"?>
  <result>
  <lex>dväljas..1</lex>
      <fm>bo..1</fm>
      <fp>PRIM..1</fp>
      <mfs></mfs>
      <pfs></pfs>
      <ls>
      <l>dväljas..vb.1</l>
      </ls>
  </result>
  '''
# ---
# name: TestLidRoutes.test_xml_valid_input_returns_200[dv\xe4ljas..vb.1]
  '''
  <?xml version="1.0" encoding="UTF-8"?>
  <result>
  <gf>dväljas</gf>
      <p>vb_vs_dväljas</p>
      <ls>
      <l>dväljas..1</l>
      </ls>
  
  </result>
  '''
# ---
